src/path.js

```diff
@@ -0,0 +1,37 @@
+import { properties } from "./children.js";
+
+/**
+ * Find a path from an ancestor node to a descendant node.
+ * @param {object} ancestor the ancestor node to start the search from
+ * @param {object} descendant the descendant node to search for
+ * @returns {(string | number)[] | null} an array of keys to traverse to get from the ancestor to the descendant, or null if no path exists
+ */
+export default function path (ancestor, descendant) {
+	// Found descendant
+	if (ancestor === descendant) {
+		return [];
+	}
+
+	// Case where ancestor is an array for an array expression or compound etc
+	if (Array.isArray(ancestor)) {
+		for (let i = 0; i < ancestor.length; i++) {
+			const childPath = path(ancestor[i], descendant);
+			if (childPath) {
+				return [i].concat(childPath);
+			}
+		}
+		return null;
+	}
+
+	const childProperties = properties[ancestor.type] ?? [];
+
+	for (const property of childProperties) {
+		const childPath = path(ancestor[property], descendant);
+		if (childPath) {
+			return [property].concat(childPath);
+		}
+	}
```

Comment:
This seems like two blocks of code doing largely the same thing.
Can we unify them? Array indices are properties after all. We could trivially unify them by using `childProperties = Array(ancestor.length).fill(0).map((_, i) => i);` for arrays, but that's a bit clumsy and a bit wasteful.

Reply:
I can't think of any other way to unify them without doing the trick you mentioned; I tried to think of one while implementing it, but couldn't think of one that wouldn't waste some performance